core/evm.go

```diff
@@ -35,49 +35,65 @@ type ChainContext interface {
 }
 
-// NewEVMContext creates a new context for use in the EVM.
-func NewEVMContext(msg Message, header *types.Header, chain ChainContext, author *common.Address) vm.Context {
+func NewBlockContext(header *types.Header, chain ChainContext, author *common.Address) *vm.BlockContext {
 	// If we don't have an explicit author (i.e. not mining), extract from the header
 	var beneficiary common.Address
 	if author == nil {
 		beneficiary, _ = chain.Engine().Author(header) // Ignore error, we're past header validation
 	} else {
 		beneficiary = *author
 	}
+	return &vm.BlockContext{
+		GetHash:      GetHashFn(header, chain),
+		CanTransfer:  CanTransfer,
+		Transfer:     Transfer,
+		Coinbase:     beneficiary,
+		BlockNumber:  new(big.Int).Set(header.Number),
+		Time:         new(big.Int).Set(header.Time),
+		Difficulty:   new(big.Int).Set(header.Difficulty),
+		GasLimit:     header.GasLimit,
+
+	}
+}
+
+// NewEVMContext creates a new context for use in the EVM.
+func NewEVMContext(msg Message, blockContext *vm.BlockContext) vm.Context {
 	return vm.Context{
-		CanTransfer: CanTransfer,
-		Transfer:    Transfer,
-		GetHash:     GetHashFn(header, chain),
-		Origin:      msg.From(),
-		Coinbase:    beneficiary,
-		BlockNumber: new(big.Int).Set(header.Number),
-		Time:        new(big.Int).Set(header.Time),
-		Difficulty:  new(big.Int).Set(header.Difficulty),
-		GasLimit:    header.GasLimit,
-		GasPrice:    new(big.Int).Set(msg.GasPrice()),
+		BlockContext: blockContext,
+		Origin:       msg.From(),
+		GasPrice:     new(big.Int).Set(msg.GasPrice()),
 	}
 }
 
 // GetHashFn returns a GetHashFunc which retrieves header hashes by number
 func GetHashFn(ref *types.Header, chain ChainContext) func(n uint64) common.Hash {
-	var cache map[uint64]common.Hash
+
+	// Cache will initially contain [refHash.parent],
+	// Then fill up with [refHash.p, refHash.pp, refHash.ppp, ...]
+	var cache []common.Hash
 
 	return func(n uint64) common.Hash {
 		// If there's no hash cache yet, make one
-		if cache == nil {
-			cache = map[uint64]common.Hash{
-				ref.Number.Uint64() - 1: ref.ParentHash,
-			}
+		if len(cache) == 0 {
+			cache = append(cache, ref.ParentHash)
 		}
-		// Try to fulfill the request from the cache
-		if hash, ok := cache[n]; ok {
-			return hash
+		if idx := ref.Number.Uint64() - n - 1; idx < uint64(len(cache)) {
+			return cache[idx]
 		}
-		// Not cached, iterate the blocks and cache the hashes
-		for header := chain.GetHeader(ref.ParentHash, ref.Number.Uint64()-1); header != nil; header = chain.GetHeader(header.ParentHash, header.Number.Uint64()-1) {
-			cache[header.Number.Uint64()-1] = header.ParentHash
-			if n == header.Number.Uint64()-1 {
-				return header.ParentHash
+		// No luck in the cache, but we can start iterating from the last element we already know
+		lastKnownHash := cache[len(cache)-1]
+		lastKnownNumber := ref.Number.Uint64() - uint64(len(cache))
+
+		for {
+			header := chain.GetHeader(lastKnownHash, lastKnownNumber)
```

Comment:
It seems like the existing `GetHeader` on the `ChainContext` is overkill. The access should be limited to what's needed and there would be a `GetCanonicalHash(uint64) common.Hash` method, which would utilize a similarly named method on `BlockChain`. Then the `BlockChain` itself would contain the cache (as mentioned in another comment).

Reply:
Yeah, might be a bit better to not get everything, I'll look into that. However, it's not only canonical hashes. We need to be able to evaluate/execute both sidechains as well as new blocks being mined, 

I do have, and still am, pondering other ways to 'cheat' by storing canon hashes somewhere, but it might be a bit messy to deal with both reorgs and sidechains. So I opted to improve the existing schema here, and see how much better it performs with these two optimizations.

Reply:
Makes sense. Just wanted to point that out.